lib/cudnn/nnlib.jl

```diff
@@ -76,11 +104,17 @@ function conv_bias_act!(y::CuArray{T}, x::CuArray{T}, w::CuArray{T}, cdims::Dens
   # only relu and identity are supported
   if σ == NNlib.relu
     if algo < 0
-      # algo = UInt32(cudnnGetConvolutionForwardAlgorithm(fix1d(y), fix1d(x), fix1d(w), fix1d(cdims))) # will be removed in cuDNN 8
-      returnedAlgoCount, perfResults = cudnnGetConvolutionForwardAlgorithm_v7(fix1d(y), fix1d(x), fix1d(w), fix1d(cdims))
-      # returnedAlgoCount, perfResults = cudnnFindConvolutionForwardAlgorithm(fix1d(y), fix1d(x), fix1d(w), fix1d(cdims))
-      # returnedAlgoCount, perfResults = cudnnFindConvolutionForwardAlgorithmEx(fix1d(y), fix1d(x), fix1d(w), fix1d(cdims))
-      algo = UInt32(perfResults[1].algo)
+      global conv_forward
+      key = (T, size(x)[end], cdims)
+      algo = get(conv_forward, key, -1)
+      if algo < 0 # not in conv_forward
+        # algo = UInt32(cudnnGetConvolutionForwardAlgorithm(fix1d(y), fix1d(x), fix1d(w), fix1d(cdims))) # will be removed in cuDNN 8
+        # returnedAlgoCount, perfResults = cudnnGetConvolutionForwardAlgorithm_v7(fix1d(y), fix1d(x), fix1d(w), fix1d(cdims))
+        # returnedAlgoCount, perfResults = cudnnFindConvolutionForwardAlgorithm(fix1d(y), fix1d(x), fix1d(w), fix1d(cdims))
+        returnedAlgoCount, perfResults = cudnnFindConvolutionForwardAlgorithmEx(fix1d(y), fix1d(x), fix1d(w), fix1d(cdims), workspacesize=workspacesize(fix1d(x)))
+        algo = UInt32(perfResults[1].algo)
+        push!(conv_forward, key, algo)
```

Comment:
Do other properties of y/x/w not influence the selected algorithm?

Reply:
I have added the strides of x/y/w in https://github.com/JuliaGPU/CUDA.jl/pull/321/commits/41c7fe5823917477011d7fe253e6fccd66e3b2d2.
The spatial size of x/y/w, the padding, stride, dilation, ... are all inside the `cdims` or can be calculated using only depending on the elements in cdims.
The data type and batch size are not in cdims, so I added those already.